File: output_20260320_065752/claude-opus/java/Linklist/Spouse.java
class Person {

/*@
  predicate inv(Person this) = this.spouse |-> ?s &*& (s == null || s.spouse == this);
@*/

    protected Person spouse;

    public void spouse_symm()
    //@ requires inv(this);
    //@ ensures inv(this);
    {
        //@ open inv(this);
        //@ if (spouse != null) { open inv(spouse); }
        //@ if (spouse != null) { assert spouse.spouse == this; }
        //@ if (spouse != null) { close inv(spouse); }
        //@ close inv(this);
    }

    public Person()
    //@ requires true;
    //@ ensures inv(this);
    {
        //@ close inv(this);
    }

    public Person getSpouse()
    //@ requires inv(this);
    //@ ensures inv(this) &*& result == this.spouse;
    {
        //@ open inv(this);
        Person r = spouse;
        //@ close inv(this);
        return r;
    }

    protected void setSpouse(Person other)
    //@ requires inv(this) &*& inv(other);
    //@ ensures inv(this) &*& inv(other);
    {
        //@ open inv(this);
        //@ open inv(other);
        spouse = other;
        other.spouse = this;
        //@ close inv(other);
        //@ close inv(this);
    }

    protected void clearSpouse()
    //@ requires inv(this);
    //@ ensures inv(this);
    {
        //@ open inv(this);
        if (spouse != null) {
            //@ open inv(spouse);
            spouse.spouse = null;
            //@ close inv(spouse);
        }
        spouse = null;
        //@ close inv(this);
    }

    void marry(Person other)
    //@ requires inv(this) &*& inv(other);
    //@ ensures inv(this) &*& inv(other) &*& this.spouse == other &*& other.spouse == this;
    {
        other.setSpouse(this);
    }

    void divorce()
    //@ requires inv(this);
    //@ ensures inv(this);
    {
        //@ open inv(this);
        if (spouse != null) {
            spouse.clearSpouse();
        }
        //@ close inv(this);
    }

}

class Program {

    public static void main(String[] args)
    //@ requires true;
    //@ ensures true;
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}

File: output_at_5/gpt52/sample_3/c/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

struct person {
  char* name;
  struct person* spouse;
};

/*@
predicate person(struct person *p; struct person *s) =
  p != 0 &*& malloc_block_person(p) &*& p->name |-> _ &*& p->spouse |-> s;
@*/

struct person *create_person()
  //@ requires true;
  //@ ensures person(result, 0);
  
  
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  p->spouse = 0;
  //@ close person(p, 0);
  
  return p;
}

void marry(struct person *this, struct person *other)
  //@ requires person(this, 0) &*& person(other, 0) &*& this != other;
  //@ ensures person(this, other) &*& person(other, this);
  
  
{
  //@ open person(this, 0);
  //@ open person(other, 0);
  
  this->spouse = other;
  other->spouse = this;
  
  //@ close person(this, other);
  //@ close person(other, this);
  
}

struct person* person_get_spouse(struct person* this)
  //@ requires person(this, ?s);
  //@ ensures person(this, s) &*& result == s;
  
  
{
  //@ open person(this, s);
  
  struct person* res = this->spouse;
  
  //@ close person(this, s);
  return res;
  
}

void divorce(struct person* this)
  //@ requires person(this, ?other) &*& other != 0 &*& person(other, this);
  //@ ensures person(this, 0) &*& person(other, 0);
  
  
{
  //@ open person(this, other);
  //@ open person(other, this);
  
  this->spouse->spouse = 0;
  this->spouse = 0;
  
  //@ close person(other, 0);
  //@ close person(this, 0);
  
}

void die(struct person *this)
  //@ requires person(this, ?s) &*& (s == 0 ? true : person(s, this));
  //@ ensures s == 0 ? true : person(s, 0);
  
  
{
  //@ open person(this, s);
  
  if(this->spouse != 0) {
    //@ open person(s, this);
    
    this->spouse->spouse = 0;
    
    //@ close person(s, 0);
  }
  free(this); 
}

int main() 
  //@ requires true;
  //@ ensures true;
  
  
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  //@ open person(bob, eve);
  //@ open person(eve, bob);
  die(bob);
  //@ open person(eve, 0);
  die(eve);
  return 0;
}